VitalHealthConnect/src/main/java/io/tryvital/vitalhealthconnect/workers/ResourceSyncWorker.kt

```diff
@@ -0,0 +1,177 @@
+package io.tryvital.vitalhealthconnect.workers
+
+import android.content.Context
+import android.content.SharedPreferences
+import androidx.work.CoroutineWorker
+import androidx.work.Data
+import androidx.work.WorkerParameters
+import com.squareup.moshi.JsonClass
+import com.squareup.moshi.Moshi
+import com.squareup.moshi.adapters.PolymorphicJsonAdapterFactory
+import com.squareup.moshi.adapters.Rfc3339DateJsonAdapter
+import com.squareup.moshi.kotlin.reflect.KotlinJsonAdapterFactory
+import io.tryvital.client.Environment
+import io.tryvital.client.Region
+import io.tryvital.client.VitalClient
+import io.tryvital.client.utils.VitalLogger
+import io.tryvital.vitalhealthconnect.HealthConnectClientProvider
+import io.tryvital.vitalhealthconnect.model.VitalResource
+import io.tryvital.vitalhealthconnect.records.*
+import io.tryvital.vitalhealthconnect.records.HealthConnectRecordAggregator
+import io.tryvital.vitalhealthconnect.records.HealthConnectRecordProcessor
+import io.tryvital.vitalhealthconnect.records.HealthConnectRecordReader
+import kotlinx.coroutines.delay
+import java.util.*
+
+internal val moshi by lazy {
+    Moshi.Builder()
+        .add(Date::class.java, Rfc3339DateJsonAdapter())
+        .add(ResourceSyncState.adapterFactory)
+        .addLast(KotlinJsonAdapterFactory())
+        .build()
+}
+
+data class ResourceSyncWorkerInput(
+    val resource: VitalResource,
+
+    // TODO: Remove after SDK singletonization
+    val region: Region,
+    val environment: Environment,
+    val apiKey: String,
+) {
+    fun toData(): Data = Data.Builder().run {
+        putString("resource", resource.toString())
+        putString("environment", environment.name)
+        putString("region", region.name)
+        putString("apiKey", apiKey)
+        build()
+    }
+
+    companion object {
+        fun fromData(data: Data) = ResourceSyncWorkerInput(
+            resource = VitalResource.valueOf(
+                data.getString("resource") ?: throw IllegalArgumentException("Missing resource")
+            ),
+            environment = Environment.valueOf(
+                data.getString("environment") ?: throw IllegalArgumentException("Missing environment")
+            ),
+            region = Region.valueOf(
+                data.getString("region") ?: throw IllegalArgumentException("Missing region")
+            ),
+            apiKey = data.getString("apiKey") ?: throw IllegalArgumentException("Missing API key"),
+        )
+    }
```

Comment:
why do we need this conversion?

Reply:
We manage the work using Android WorkManager framework, which requires input data to be serializable.

Reply:
WorkManager is the platform standard to manage any persistent work (background work included), and down the line it would enable various options to run our Workers as foreground services (that should not be subject to Health Connect restrictions AFAIK).

Reply:
https://github.com/tryVital/vital-android/blob/5d495c90bddc45c4bbe953d348ba12f972b5a388/VitalHealthConnect/src/main/java/io/tryvital/vitalhealthconnect/workers/UploadAllDataWorker.kt#L38-L52